src/gradualizer.erl

```diff
@@ -83,23 +78,234 @@ type_check_dir(Dir, Opts) ->
     type_check_files(Files, []).
 
 %% @doc Type check a source or beam
-%% (Option `print_file' is implicitely true)
--spec type_check_files([file:filename()], options()) -> ok | nok.
+-spec type_check_files([file:filename()], typechecker:options()) -> [{file:filename(), [typechecker:type_error()]}].
 type_check_files(Files, Opts) ->
     StopOnFirstError = proplists:get_bool(stop_on_first_error, Opts),
     lists:foldl(
-        fun(File, Res) when Res =:= ok;
+        fun(File, Errors) when Errors =:= [];
                             not StopOnFirstError ->
-                case type_check_file(File, [print_file|Opts]) of
-                    ok -> Res;
-                    nok -> nok
+                case type_check_file(File, Opts) of
+                    [] -> Errors;
+                    NewErrors -> [{File, NewErrors} | Errors]
                 end;
-            (_, nok) ->
-                nok
-        end, ok, Files).
+            (_, Errors) ->
+                Errors
+        end, [], Files).
+
+-spec format_error(typechecker:type_error()) -> string().
+format_error({type_error, LINE, Cause, Mismatch}) ->
+    format_error_(erl_anno:line(LINE), Cause, Mismatch).
+
+-spec format_error_(non_neg_integer(), any(), any()) -> string().
+format_error_(LINE, {call, Func, Arity}, undef) ->
+    io_lib:format("Call to undefined function ~p/~p on line ~p~n",
+        [Func, Arity, LINE]);
+format_error_(LINE, {call, Func, Arity, Module}, undef) ->
+    io_lib:format("Call to undefined function ~p:~p/~p on line ~p~n",
+        [Module, Func, Arity, LINE]);
+format_error_(LINE, {record, RecName, Module}, undef) ->
+    io_lib:format("Undefined record ~p:~p on line ~p~n",
+        [Module, RecName, LINE]);
+format_error_(LINE, {record, RecName}, undef) ->
+    io_lib:format("Undefined record ~p on line ~p~n",
+        [RecName, LINE]);
+format_error_(LINE, {remote_type, Name, Arity, Module}, undef) ->
+    io_lib:format("Undefined remote type ~p:~p/~p on line ~p~n",
+        [Module, Name, Arity, LINE]);
+format_error_(LINE, {user_type, Name, Arity, Module}, undef) ->
+    io_lib:format("Undefined user type ~p:~p/~p on line ~p~n",
+        [Module, Name, Arity, LINE]);
+format_error_(LINE, {user_type, Name, Arity}, undef) ->
+    io_lib:format("Undefined user type ~p/~p on line ~p~n",
+        [Name, Arity, LINE]);
+format_error_(LINE, {remote_type, Name, Arity, Module}, not_exported) ->
+    io_lib:format("The type ~s:~s/~p on line ~p is not exported~n",
+        [erl_pp:expr(Module), erl_pp:expr(Name), Arity, LINE]);
+format_error_(LINE, {var, Var, VarTy}, {type_mismatch, Ty}) ->
+    io_lib:format("The variable ~p on line ~p has type ~s "
+        "but is expected to have type ~s~n",
+        [Var, LINE, typelib:pp_type(VarTy), typelib:pp_type(Ty)]);
+format_error_(LINE, {char, Char}, {type_mismatch, Ty}) ->
+    io_lib:format("The character ~p on line ~p does not have type ~s~n",
+        [erl_pp:expr(Char), LINE, typelib:pp_type(Ty)]);
```

Comment:
It might be possible to merge the cases but that would require thinking - which I did not really apply here. Same goes for the wrapper - the translation is mostly mechanical. Could the erl_pp representation change? I'd like to keep using `erl_pp:expr(Char)` without knowing the internal representation.

Reply:
Sure, that's what I mean. `erl_pp:expr/1` can be used for any expression (AST).
```Erlang
format_error({expr_type_mismatch, Expr, Ty}) ->
    io_lib:format("The expression ~p on line ~p does not have type ~s~n",
        [erl_pp:expr(Expr), get_line(Expr), typelib:pp_type(Ty)]);
```